Session Affinity design (#235)
microsoft/reverse-proxy: docs/designs/session-affinity.md

# Session Affinity

## Concept
Session affinity is a mechanism to bind (affinitize) a causally related request sequence to the destination handled the first request when the load is balanced among several destinations. It is useful in scenarios where the most requests in a sequence work with the same data and the cost of data access differs for different nodes (destinations) handling requests. The most common example is a transient caching (e.g. in-memory) where the first request fetches data from a slower persistent storage into a fast local cache and the others work only with the cached data thus increasing throughput.

## Affinity Key
Request to destination affinity is established via the affinity key identifying the target destination. That key can be stored on different request parts depending on the given session affinity implementation, but each request cannot have more than one such key. The exact key semantics is implementation dependent, in example the both of built-in `CookieSessionAffinityProvider` and `CustomHeaderAffinityProvider` are currently use `DestinationId` as the affinity key.

> The current design doesn't require for key to uniquely identify the single affinitized destination, it's allowed to establish affinity to a destination group. In such case, the exact destination to handle the given request will be determined by the load balancer.

Reviewer: ```suggestion
The current design doesn't require keys to uniquely identify the single affinitized destination. It's allowed to establish affinity to a destination group. In this case, the exact destination to handle the given request will be determined by the load balancer.
```

There are two built-in affinity modes differing only in how the affinity key is stored on a request.
1. `Cookie` - stores the key as a cookie. It expects the request's key to be delivered as a cookie with configured name and sets the same cookie with `Set-Cookie` header on the first response in an affinitized sequence. This mode is implemented by `CookieSessionAffinityProvider`. Cookie's options (e.g. name) can be configured via `CookieSessionAffinityProviderOptions`. The default name is `.Microsoft.ReverseProxy.Affinity`

2. `CustomHeader` - stores the key on a header. It expects the request's key to be delivered in a customer header with configured name and sets the same header on the first response in an affinitized sequence. This mode is implemented by `CustomHeaderSessionAffinityProvider`. Header name can be set by the setting `CustomHeaderName` in `SessionAffinityOptions.Settings` as shown in the example below. The default name is `X-Microsoft-Proxy-Affinity`.

## Affinity failure policy

> If the affinity key cannot be decoded or no target destination found, it's considered as a failure, and an affinity failure policy is called to handle it. The policy has the full access to `HttpContext` and can send response to the client by itself. It returns a boolean value indicating whether the request processing can proceed down the pipeline or must be terminated.

Reviewer 2: ```suggestion
If the affinity key cannot be decoded or no healthy destination found it's considered as a failure and an affinity failure policy is called to handle it. The policy has the full access to `HttpContext` and can send response to the client by itself. It returns a boolean value indicating whether the request processing can proceed down the pipeline or must be terminated.
```

> There are two built-in failure policy:

Reviewer 2: ```suggestion
There are two built-in failure policies.  The default is `Redistribute`.
```

> 1. `Redistribute` - tries to establish a new affinity to one of available healthy destinations by skipping the affinity lookup step and passing all healthy destination to the load balance the same way it is done for a request without any affinity. Request processing continues. Implemented by `RedistributeAffinityFailurePolicy`.

Reviewer 2: ```suggestion
1. `Redistribute` - tries to establish a new affinity to one of available healthy destinations by skipping the affinity lookup step and passing all healthy destination to the load balancer the same way it is done for a request without any affinity. Request processing continues. This is implemented by `RedistributeAffinityFailurePolicy`.
```

> 2. `Return503Error` - send a `503` response back to the client. Request processing is terminated. Implemented by `Return503ErrorAffinityFailurePolicy`

Reviewer 2: ```suggestion
2. `Return503Error` - sends a `503` response back to the client and request processing is terminated. This is implemented by `Return503ErrorAffinityFailurePolicy`
```

## Request pipeline

> Session affinity mechanism is implemented by the services (mentioned above) and the two following middlewares:

Reviewer 2: ```suggestion
The session affinity mechanisms are implemented by the services (mentioned above) and the two following middleware:
```
